common/src/main/java/io/netty/util/NettyRuntime.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util;
+
+import io.netty.util.internal.SystemPropertyUtil;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A utility class for wrapping calls to {@link Runtime}.
+ */
+public final class NettyRuntime {
+
+    private static final AtomicInteger availableProcessors = new AtomicInteger();
+
+    /**
+     * No public constructor to prevent instances from being created.
+     */
+    private NettyRuntime() {
+    }
+
+    /**
+     * Set the number of available processors.
+     *
+     * @param availableProcessors the number of available processors
+     * @throws IllegalArgumentException if the specified number of available processors is non-positive
+     * @throws IllegalStateException    if the number of available processors is already configured
+     */
+    public static void setAvailableProcessors(final int availableProcessors) {
+        if (availableProcessors <= 0) {
+            throw new IllegalArgumentException(
+                    "availableProcessors must be positive but was [" + availableProcessors + "]");
+        }
+        if (!NettyRuntime.availableProcessors.compareAndSet(0, availableProcessors)) {
+            throw new IllegalStateException(
+                    "availableProcessors is already set to [" + NettyRuntime.availableProcessors.get() + "], " +
+                            "rejecting [" + availableProcessors + "]");
+        }
+    }
+
+    /**
+     * Get the configured number of available processors. The default is {@link Runtime#availableProcessors()}. This
+     * can be overridden by setting the system property "io.netty.availableProcessors" or by invoking
+     * {@link #setAvailableProcessors(int)} before any calls to {@link #availableProcessors()}.
+     *
+     * @return the configured number of available processors
+     */
+    @SuppressForbidden(reason = "to obtain default number of available processors")
+    public static int availableProcessors() {
+        if (NettyRuntime.availableProcessors.get() == 0) {
+            final int availableProcessors =
+                    SystemPropertyUtil.getInt(
+                            "io.netty.availableProcessors",
+                            Runtime.getRuntime().availableProcessors());
+            setAvailableProcessors(availableProcessors);
```

Comment:
> I still don't understand why you think this can never happen. How can you guarantee this ?

@normanmaurer I never said it can't happen (notice I said "shouldn't", not "can't"). I've acknowledged from the beginning it can happen and I coded this with the possibility in mind. To be clear, a get race is possible, we're just arguing about how to handle it. You want to allow it, I want to forbid it (in the sense that if an application developer causes a get race to exist, it should be treated as an application bug and dealt with there). I want it this way because it's simple. I don't want it your way because your way permits get and set races in the sense that if an application allows a race to exist between a get and set, their application can appear to run fine if the set executes before the get until one day an evil scheduler runs the get before the set and then the application explodes. My way has a simple model: do not permit a race in your application code. My way is not restrictive because there's no practical situation in which obtaining this value should be racy. Look at the uses within Netty. Again, look at the example of `NioEventLoopGroup` I gave above. Also, a get race is not as innocent as you make it sound because the system property can be set in code; so a get race like you want to allow could have one thread win before the system property is set on another thread that tries to get the value to set it lazily yet sometimes that thread wins and the application ends up with different values depending on the winner. Let's have none of this, with a very clear and easy to reason about design. Let's not design this for a situation that is easily avoided and has no practical advantage in allowing.